File: Codes/Kan Zheng/C++/sort/75. Sort Colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int n=nums.size();
        int i=0;
        while(n--)
        {
            if(nums[i]==0)
            {
                nums.insert(nums.begin(),nums[i]);
                nums.erase(nums.begin()+i+1);
                i++;
            }
            else if (nums[i]==2)
            {
                nums.insert(nums.end(),nums[i]);
                nums.erase(nums.begin()+i);
            }
            else
                i++;
        }
    }
};

File: Codes/Bingran Li/Sort/75. Sort Colors.java
public class Solution {
    public void sortColors(int[] nums) {
        // solution 1: two pass algorithm---counting sort!
        /*
        int count0=0,count1=0,count2=0;
        for(int i:nums){
            if(i==0) count0++;
            if(i==1) count1++;
            if(i==2) count2++;
        }
        for(int i=0;i<nums.length;i++){
            if(i<count0)     nums[i]=0;
            else if(i<count1+count0) nums[i] = 1;
            else nums[i]=2;
        }
        */
        // solution 2: one pass algorithm----trickey
        
        int second =nums.length-1;
        int zero = 0;
        for(int i=0;i<=second;i++){
             
            while(nums[i]==2 && i<second ){
                int tem = nums[second];
                nums[second] = 2;
                nums[i] = tem;
                second--;
            }
            while(nums[i]==0 && i>zero ){
                int tem=nums[zero];
                nums[zero]=0;
                nums[i]=tem;
                zero++;
            }
           
        }
            
        
        
    }
}